fluent/src/bundle.js

```diff
@@ -170,74 +153,124 @@ export default class FluentBundle {
       if (id.startsWith("-")) {
         // Identifiers starting with a dash (-) define terms. Terms are private
         // and cannot be retrieved from FluentBundle.
         if (allowOverrides === false && this._terms.has(id)) {
           errors.push(`Attempt to override an existing term: "${id}"`);
           continue;
         }
-        this._terms.set(id, value);
+        this._terms.set(id, entry);
       } else {
         if (allowOverrides === false && this._messages.has(id)) {
           errors.push(`Attempt to override an existing message: "${id}"`);
           continue;
         }
-        this._messages.set(id, value);
+        this._messages.set(id, entry);
       }
     }
 
     return errors;
   }
 
   /**
-   * Format a message to a string or null.
+   * Format a message to a string.
+   *
+   * Find a message by `id` in the bundle and format it into a string.  If the
+   * message has no value, return null.
    *
-   * Format a raw `message` from the bundle into a string (or a null if it has
-   * a null value).  `args` will be used to resolve references to variables
-   * passed as arguments to the translation.
+   * `args` can be an object or `null`. If it's an object, it will be used to
+   * resolve references to variables passed as arguments to the translation.
    *
    * In case of errors `format` will try to salvage as much of the translation
    * as possible and will still return a string.  For performance reasons, the
    * encountered errors are not returned but instead are appended to the
    * `errors` array passed as the third argument.
    *
-   *     const errors = [];
-   *     bundle.addMessages('hello = Hello, { $name }!');
-   *     const hello = bundle.getMessage('hello');
-   *     bundle.format(hello, { name: 'Jane' }, errors);
+   *     let errors = [];
+   *     bundle.addMessages("hello = Hello, {$name}!");
    *
-   *     // Returns 'Hello, Jane!' and `errors` is empty.
+   *     bundle.format("hello", {name: "Jane"}, errors);
+   *     // → "Hello, Jane!" and `errors` is empty.
    *
-   *     bundle.format(hello, undefined, errors);
+   *     bundle.format("hello", undefined, errors);
+   *     // → "Hello, name!" and `errors` is now:
+   *     // [<ReferenceError: Unknown variable: name>]
    *
-   *     // Returns 'Hello, name!' and `errors` is now:
-   *
-   *     [<ReferenceError: Unknown variable: name>]
-   *
-   * @param   {Object | string}    message
-   * @param   {Object | undefined} args
-   * @param   {Array}              errors
+   * @param   {string} id
+   * @param   {?Object} args
+   * @param   {?Array} errors
    * @returns {?string}
    */
-  format(message, args, errors) {
-    // optimize entities which are simple strings with no attributes
-    if (typeof message === "string") {
-      return this._transform(message);
+  format(id, args, errors = []) {
+    if (!this._messages.has(id)) {
+      errors.push(`Message not found: "${id}"`);
+      return undefined;
```

Comment:
Why `undefined` rather than `null`? Is the user expected to always check if the message exists before trying to format it? I thought `format("non-existing") -> null` is a reasonable API use and not an error?

Reply:
I decided to used `undefined` here and when formatting missing attributes in order to tell missing values from null values apart. Consider:

```properties
foo =
    .attr = Attr
```

- `format("foo")` → `null`,
- `format("bar")` → `undefined`.

Do you think they should both return `null` instead?

Reply:
That's an interesting question. I'm torn. For me the question boils down to - should attempting to format a message that doesn't existing in the bundle be an error (which means, you should always call `hasMessage` before you call format)?

Reply:
I think the bindings should always check, yes. They should also use `formatValue` and `formatAttribute`, rather than `format`. `format` is intended for more imperative approaches. The fact that it returns `undefined` on missing messages was intended to be consistent with `formatAttribute` returning `undefined` on missing attributes.

It also works well semantically, I think. The difference between `null` and `undefined` maps nicely to Fluent's notions of value-less messages vs. missing messages. Attempting to format a missing message should result in an error being reported. As a developer, there's little incentive for me to intentionally format missing messages :) So it's likely an error scenario, which I'd like to be able to react to.